beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -2616,21 +2590,29 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
                 .map_err(BeaconChainError::from)?;
         }
 
-        let mut fork_choice = self.fork_choice.write();
+        let mut fork_choice = self.canonical_head.fork_choice_write_lock();
 
         // Do not import a block that doesn't descend from the finalized root.
-        let signed_block =
-            check_block_is_finalized_descendant::<T, _>(signed_block, &fork_choice, &self.store)?;
-        let (block, block_signature) = signed_block.clone().deconstruct();
+        check_block_is_finalized_descendant(self, &fork_choice, &signed_block)?;
```

Comment:
```suggestion
        check_block_is_finalized_descendant(self, &self.canonical_head.fork_choice_read_lock(), &signed_block)?;
```

Github won't let me suggest I'm trying to suggest after the
```rust
if let Some(wss_checkpoint) = self.config.weak_subjectivity_checkpoint {
    ...
    let current_head_finalized_checkpoint =
        self.canonical_head.cached_head().finalized_checkpoint();
    ...
}
```
block below, you insert:
```rust
let mut fork_choice = self.canonical_head.fork_choice_write_lock();
```

For 2 reasons:
1. Lock contention *slightly* reduced
2. Inside the `if` statement you call `self.canonical_head.cached_head()` which acquires the `cached_head` read lock. Now from what I can see, this isn't an issue right now. But if at some point in the future, a piece of code acquires the `cached_head` write lock, and then simultaneously tries to acquire *either* of the `fork_choice` read or write locks, we can have a deadlock.

Reply:
Nice one, this is a great pickup :1st_place_medal: :1st_place_medal:  I ended up moving the WS check to *before* we start doing things with fork choice. I believe it's functionally equivalent and it also avoids us having to "bounce" the locks (i.e., taking a read-lock, dropping it, then taking a write-lock).